File: shared/video-protocol/ts/src/protocol.ts
/**
 * Constants for the UDP video streaming protocol.
 * See: shared/video-protocol/UDP_Frame_Format_Documentation.md
 *
 * NOTE: Web browsers cannot use raw UDP sockets.
 * This package targets Node.js — use it in a WebSocket bridge that forwards
 * UDP packets from the Raspberry Pi to browser clients.
 */

export const SERVER_PORT = 9999;
export const CHUNK_PAYLOAD_SIZE = 1200;
export const CLIENT_TIMEOUT_MS = 30_000;
export const KEEPALIVE_INTERVAL_MS = 5_000;

// Packet markers (UTF-8 ASCII)
export const FRAME_START_MARKER = Buffer.from("FRAME_START");  // 11 bytes
export const CHUNK_MARKER       = Buffer.from("CHUNK");        //  5 bytes

// Control messages
export const MSG_REGISTER_CLIENT = Buffer.from("REGISTER_CLIENT");
export const MSG_REGISTERED      = Buffer.from("REGISTERED");
export const MSG_KEEPALIVE       = Buffer.from("KEEPALIVE");
export const MSG_DISCONNECT      = Buffer.from("DISCONNECT");

/**
 * Wire format for header integer fields.
 *
 * Raspberry Pi 64-bit (primary):  struct.pack("LLL") → BigUInt64LE
 *   FRAME_START total: 35 bytes  (11 + 3×8)
 *   CHUNK header:      21 bytes  ( 5 + 2×8)
 *
 * Legacy 32-bit (fallback):       struct.pack("III") → UInt32LE
 *   FRAME_START total: 23 bytes  (11 + 3×4)
 *   CHUNK header:      13 bytes  ( 5 + 2×4)
 */
export type HeaderFormat = "uint64" | "uint32";

export function detectFrameStartFormat(buf: Buffer): HeaderFormat | null {
  if (buf.length === 35) return "uint64";
  if (buf.length >= 23) return "uint32";
  return null;
}

export function detectChunkFormat(buf: Buffer): HeaderFormat | null {
  if (buf.length >= 21) return "uint64";  // 5 + 16 + ≥1 payload byte
  if (buf.length >= 13) return "uint32";  // 5 +  8 + ≥1 payload byte (or just header)
  return null;
}

File: shared/video-protocol/swift/Sources/VideoProtocol/Protocol.swift
import Foundation

/// Namespace for the wrack UDP video streaming protocol constants.
///
/// Wire format overview:
///
/// **Control messages (client → server)**
/// | Message            | Bytes               |
/// |--------------------|---------------------|
/// | `REGISTER_CLIENT`  | UTF-8 string, 15 B  |
/// | `KEEPALIVE`        | UTF-8 string,  9 B  |
/// | `DISCONNECT`       | UTF-8 string, 10 B  |
///
/// **Control messages (server → client)**
/// | Message            | Bytes               |
/// |--------------------|---------------------|
/// | `REGISTERED`       | UTF-8 string, 10 B  |
///
/// **Data packets (server → client)**
///
/// FRAME_START (primary 64-bit, 35 bytes):
///   `FRAME_START` (11 B) | frame_id UInt64 LE | frame_size UInt64 LE | chunk_count UInt64 LE
///
/// FRAME_START (fallback 32-bit, 23 bytes):
///   `FRAME_START` (11 B) | frame_id UInt32 LE | frame_size UInt32 LE | chunk_count UInt32 LE
///
/// CHUNK (primary 64-bit, 21 B header + payload):
///   `CHUNK` (5 B) | frame_id UInt64 LE | chunk_index UInt64 LE | payload (≤ chunkPayloadSize)
///
/// CHUNK (fallback 32-bit, 13 B header + payload):
///   `CHUNK` (5 B) | frame_id UInt32 LE | chunk_index UInt32 LE | payload (≤ chunkPayloadSize)
///
/// See: `shared/video-protocol/UDP_Frame_Format_Documentation.md`
public enum VideoProtocol {

    // MARK: - Network

    /// Default server port.
    public static let serverPort: UInt16 = 9999

    // MARK: - Framing

    /// Maximum payload bytes per CHUNK datagram.
    /// Sized to avoid IP fragmentation on most networks.
    public static let chunkPayloadSize: Int = 1200

    // MARK: - Timings

    /// Seconds before an idle client is evicted by the server.
    public static let clientTimeout: TimeInterval = 30

    /// Seconds between KEEPALIVE messages sent by the client.
    public static let keepaliveInterval: TimeInterval = 5

    // MARK: - Packet markers (internal — used by Packets.swift)

    static let frameStartMarker = Data("FRAME_START".utf8)  // 11 bytes
    static let chunkMarker      = Data("CHUNK".utf8)        //  5 bytes

    // MARK: - Control messages

    /// Sent by the client to register with the server and begin receiving frames.
    public static let registerClient = Data("REGISTER_CLIENT".utf8)

    /// Sent by the server to confirm registration.
    public static let registered     = Data("REGISTERED".utf8)

    /// Sent by the client periodically to prevent server-side timeout eviction.
    public static let keepalive      = Data("KEEPALIVE".utf8)

    /// Sent by the client to gracefully deregister before closing the socket.
    public static let disconnect     = Data("DISCONNECT".utf8)
}
